### ISSUE TITLE
MarkdownHeaderTextSplitter

### ISSUE THREAD

**Issue by andyflury** [langchain4j/langchain4j#574]
langchain4j already offers several DocumentSplitters. One that is currently missing is MarkdownHeaderTextSplitter.

The original (python based) langchain project has such a [MarkdownHeaderTextSplitter](https://python.langchain.com/docs/modules/data_connection/document_transformers/markdown_header_metadata)

Would be nice to have this as well in langchain4j.

Attached an implementation for this. I used ChatGPT to translate the Python code to java (incl. the Unit Test). It might not conform to all the coding standards of the project, but it does the job and tests pass.

[MarkdownHeaderTextSplitter.zip](https://github.com/langchain4j/langchain4j/files/14108652/MarkdownHeaderTextSplitter.zip)

**Comment by langchain4j** [comment]
@andyflury thank you!

**Comment by mike-adonis** [comment]
Hey @langchain4j can i pick this up ?

**Comment by langchain4j** [comment]
@mike-adonis sure, go ahead! Thank you!

**Comment by langchain4j** [comment]
@mike-adonis did you manage to start/implement it?

**Comment by mike-adonis** [comment]
Yes I did start, just been busy with a few things I can try to complete it by the end of the weekend

**Comment by mike-adonis** [comment]
@langchain4j please find the pr for this #690
